### DISCUSSION TITLE
Deeplinks Discussion

### DISCUSSION THREAD

**Discussion by wer-mathurin** [Milad-Akarie/auto_route_library#1922] [Q&A]
@Milad-Akarie
Let say we have a host : example.com

And we want to "filter" the default open app for Android.
We can use the android:patPrefix to add for example : /myFilter

So the activity will only be invoke when the user click on a links like this : 
example.com/myFilter

But the bad news is that router will never have a route /myFilter

I would like to understand if the package will support this scenario:
user click on a link example.com/myFilter/ANYTHING

I would like to have the deeplinkBuilder method call (this is not working today)
I the works case, we can create a dummy page...to have the deeplinkbuilder called.

So I hope the question is clear enough.

**Comment by Milad-Akarie** [comment]
Hey @wer-mathurin are you saying that the deeplink builder method is never invoked?

**Comment by wer-mathurin** [comment]
Exactly

Only if the path matches what's defined in the router
Le jeu. 11 avr. 2024, 17 h 03, Milad Akarie ***@***.***> a
écrit :

**Comment by wer-mathurin** [comment]
@Milad-Akarie When looking at the code this is essentially case by:

When you receing a signal from the platform
```
  @override
  Future<bool> didPushRouteInformation(
      RouteInformation routeInformation) async {
    assert(hasListeners);
    _platformReportsNewRouteInformation(routeInformation);
    return true;
  }
```

at some point you are validating if the request have a matches and store the information in UrlState

```
  @override
  Future<UrlState> parseRouteInformation(RouteInformation routeInformation) {
    final resolvedUri = _normalize(routeInformation.uri);
    var matches = _matcher.matchUri(resolvedUri,
        includePrefixMatches: includePrefixMatches);
    return SynchronousFuture<UrlState>(
      UrlState(routeInformation.uri, matches ?? const [],
          pathState: routeInformation.state),
    );
  }
```

This information is pass then to **delegateRouteSetter** the _data_ variable
```
  _RouteSetter<T> _processParsedRouteInformation(Object? transaction, ValueGetter<_RouteSetter<T>> delegateRouteSetter) {
    return (T data) async {
      if (_currentRouterTransaction != transaction) {
        return;
      }
      await delegateRouteSetter()(data);  <=========================
      if (_currentRouterTransaction == transaction) {
        _rebuild();
      }
    };
  }
```
And then since there is no match the deepLinkBuilder will never be invoked

```
@override
  Future<void> setNewRoutePath(UrlState configuration) async {
    final topMost = controller.topMostRouter();
    if (topMost is StackRouter && topMost.hasPagelessTopRoute) {
      topMost.popUntil((route) => route.settings is Page);
    }

    if (configuration.hasSegments) {. <============ Since there is no match deeplinkBuilder is not invoked
      final platLink = PlatformDeepLink._(configuration, false);
      final resolvedLink = deepLinkBuilder == null ? platLink : await deepLinkBuilder!(platLink);
      if (rebuildStackOnDeepLink) {
        controller.popUntil((route) => false);
      }
      await _handleDeepLink(resolvedLink);
    }

    notifyListeners();
    return SynchronousFuture(null);
  }
```

My suggestion will be to change a little bit the logic, remove the if statement and send an ignoredeeplink by default when there is no segment.  This way the deeplinkbuilder will aways be called :-), but this a potential breaking change, because the users that override the deepLinkBuilder may receive calls now.

```
@override
  Future<void> setNewRoutePath(UrlState configuration) async {
    final topMost = controller.topMostRouter();
    if (topMost is StackRouter && topMost.hasPagelessTopRoute) {
      topMost.popUntil((route) => route.settings is Page);
    }

    
    final platLink = PlatformDeepLink._(configuration, false) ;
    final resolvedLink = deepLinkBuilder == null ? platLink : await deepLinkBuilder!(platLink);
    if (rebuildStackOnDeepLink) {
        controller.popUntil((route) => false);
    }
    await _handleDeepLink(resolvedLink);
    

    notifyListeners();
    return SynchronousFuture(null);
  }
```

Add some logic to _handleDeepLink to chatch that a PlatformDeeplink can be "invalid" and do nothing instead of throwing an error

**Comment by wer-mathurin** [comment]
@Milad-Akarie
Here is a PR for discussion base on my previous comment
https://github.com/Milad-Akarie/auto_route_library/pull/1923

**Comment by Milad-Akarie** [comment]
@wer-mathurin I've be been meaning to implement something like android's declarative prefix handling. In android you just declare what your path prefix is and everything is handled automatically, when the path is ready to be parsed it should be prefix-free.

What you suggested will definitely work but as you said it might introduce breaking changes, I also don't like the idea of first trying to find matches and if that fails you transform your path and run the matcher a second time.

What I have in mind is an implementation of a DeepLinkTransformer function that simple passes the raw uri and returns a transformed one before the matching starts, this way we don't effect existing implementation (if no transformer provided nothing changes), the matcher is only called once and everybody is happy

The api would look something like the following 

`appRouter.config(deeplinkTransformer: (uri){ 
 // transform and return new uri
}`


Mobile deep links are likely to have prefixes so it would be helpful to ship a prefix stripper transformer,

`deepLinkTransformer: PrefixStripper('myPrefix')`

What do you think? Also if you have better names for any of the above im happy to hear it

**Reply by wer-mathurin** [reply]
Awesome.

I can change my PR to do that if you want.  Also I would suggest that the return uri of the deekplinktransformer to be nullable.

This way it support the cases that the user just want to perform an action witch if not necessary going to go to a specific route.

This is my case :-)

**Comment by wer-mathurin** [comment]
@Milad-Akarie You can look at the [PR](wer-mathurin:feature/deeplinkbuilder_always), I did what you proposed.

Was not able to run the test and did not find any test related to deeplinking.

I'm missing the readme update and version increase of the plugin.  I don't know your release process.
